Turbo-Native copy in Micronaut using Thymeleaf (#979)
micronaut-projects/micronaut-guides: guides/micronaut-turbo-native/micronaut-turbo-native.adoc

common:header-top.adoc[]

common:requirements.adoc[]

== Download

The complete application can be downloaded from the following link.

* link:@sourceDir@.zip[Download] and unzip the source

== Getting Started

> This guide provides the Micronaut version, written in @language@, of https://github.com/hotwired/turbo-native-demo[Hotwire's Node small app]. It shows how to use https://turbo.hotwired.dev/[Turbo] with a native (iOS or Android) app.

Reviewer: ```suggestion
This guide provides the Micronaut version, written in @language@, of https://github.com/hotwired/turbo-native-demo[Hotwire's Node small app]. It demonstrates how to use https://turbo.hotwired.dev/[Turbo] with a native (iOS or Android) app.
```

common:hotwire.adoc[]

To run the Android application, install https://developer.android.com/studio[Android Studio].

To run the iOS application, install https://developer.apple.com/xcode/[XCode]

common:runapp.adoc[]

== Android Demo App

> https://github.com/hotwired/turbo-android/tree/main/demo[Clone or Download the Android Turbo repository].

Reviewer: ```suggestion
https://github.com/hotwired/turbo-android/tree/main/demo[Clone or download the Android Turbo repository].
```

> Once the Micronaut application is running, then you just need to change `demo/src/main/kotlin/dev/hotwire/turbo/demo/util/Constants.kt`. Set the BASE_URL to the url of the Micronaut Application.

Reviewer: ```suggestion
Once the Micronaut application is running, you just need to change `demo/src/main/kotlin/dev/hotwire/turbo/demo/util/Constants.kt`. Set the BASE_URL to the URL of the Micronaut application.
```

> NOTE: If you're running your Micronaut application locally without HTTPS, you'll need to adjust your `android:usesCleartextTraffic` settings in the `AndroidManifest.xml` (or use an Android Network security configuration), and target https://developer.android.com/studio/run/emulator-networking[10.0.2.2 instead of localhost].

Reviewer: ```suggestion
NOTE: If you're running your Micronaut application locally without HTTPS, you'll need to adjust your `android:usesCleartextTraffic` settings in the `AndroidManifest.xml` (or use an Android Network security configuration) and target https://developer.android.com/studio/run/emulator-networking[10.0.2.2 instead of localhost].
```

[source, kotlin]
.demo/src/main/kotlin/dev/hotwire/turbo/demo/util/Constants.kt
----
package dev.hotwire.turbo.demo.util

const val BASE_URL = "http://10.0.2.2:8080"
const val HOME_URL = BASE_URL
const val SIGN_IN_URL = "$BASE_URL/signin"
const val NUMBERS_URL = "$BASE_URL/numbers"
----

[source, xml]
.demo/src/main/AndroidManifest.xml
----
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:tools="http://schemas.android.com/tools"
    package="dev.hotwire.turbo.demo"
    xmlns:android="http://schemas.android.com/apk/res/android">

    <uses-permission android:name="android.permission.INTERNET"/>

    <application
        android:usesCleartextTraffic="true"
...
----

Once you run the app on an emulator or on a real device, you will see the results:

image::android-turbo-demo-app.png[Android demo app, 800]

== iOS Demo App

> https://github.com/hotwired/turbo-ios/tree/main/Demo[Clone or Download the demo IOS app].

Reviewer: ```suggestion
https://github.com/hotwired/turbo-ios/tree/main/Demo[Clone or download the demo IOS app].
```

> Once the Micronaut application is running, then you just need to change `Demo.current` to use `local` in `Demo/Demo.swift` to point to the local version of the demo. Edit the port to match the port used by the Micronaut application.

Reviewer: ```suggestion
Once the Micronaut application is running, point to the local version of the demo by changing `Demo.current` to use `local` in `Demo/Demo.swift`. Then edit the port to match the port used by the Micronaut application.
```

[source,swift]
----
...
struct Demo {
    ...
    static let local = URL(string: "http://localhost:8080")!
    static let turbolinks5Local = URL(string: "http://localhost:8080?turbolinks=1")!

    static var current: URL {
        local
    }
}
----

NOTE: Ensure the URL does not end with a slash.

> You will then be able to run this app locally (or on a real device), and see the results:

Reviewer: ```suggestion
You will then be able to run this app locally (or on a real device) and see the results:
```
